File: Java/1-50/CombinationSum2.java
public class CombinationSum2 {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = getCombinationSum2(candidates, 0, target);
        return res == null ? new ArrayList<List<Integer>>() : res;
    }
    public List<List<Integer>> getCombinationSum2(int[] candidates,
                                                  int start,
                                                  int target) {
        List<List<Integer>> res = new ArrayList<>();
        if (target == 0) {
            List<Integer> subRes = new ArrayList<>();
            res.add(subRes);
            return res;
        }
        if (candidates.length == start || candidates[start] > target) return null;
        int prev = -1;
        for (int i = start; i < candidates.length; i++) {
            while (i < candidates.length && candidates[i] == prev) i++;
            if (i < candidates.length) {
                prev = candidates[i];
                List<List<Integer>> combs = getCombinationSum2(candidates, i + 1, target - candidates[i]);
                if (combs != null) {
                    for (List<Integer> c : combs) {
                        List<Integer> subRes = new ArrayList<>();
                        subRes.add(candidates[i]);
                        subRes.addAll(c);
                        res.add(subRes);
                    }
                }
            }
        }
        return res;
    }
}

File: Python/1-50/CombinationSum2.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        ret = []
        def dfs(start, curr, target):
            if target <= 0 or start == len(candidates):
                if target == 0:
                    ret.append(curr)
                return
            for i in range(start, len(candidates)):
                if i == start or candidates[i] != candidates[i - 1]:
                    dfs(i + 1, curr + [candidates[i]], target - candidates[i])
        dfs(0, [], target)
        return ret
